documentation/proposals/Proposal - Silk.NET 3.X Generic Maths.md

```diff
@@ -0,0 +1,884 @@
+    static bool INumberBase<Vector2D<T>>.IsInteger(Vector2D<T> value);
+    static bool INumberBase<Vector2D<T>>.IsNaN(Vector2D<T> value);
+    static bool INumberBase<Vector2D<T>>.IsNegative(Vector2D<T> value);
+    static bool INumberBase<Vector2D<T>>.IsNegativeInfinity(Vector2D<T> value);
+    static bool INumberBase<Vector2D<T>>.IsNormal(Vector2D<T> value);
+    static bool INumberBase<Vector2D<T>>.IsOddInteger(Vector2D<T> value);
+    static bool INumberBase<Vector2D<T>>.IsPositive(Vector2D<T> value);
+    static bool INumberBase<Vector2D<T>>.IsPositiveInfinity(Vector2D<T> value);
+    static bool INumberBase<Vector2D<T>>.IsRealNumber(Vector2D<T> value);
+    static bool INumberBase<Vector2D<T>>.IsSubnormal(Vector2D<T> value);
+    static bool INumberBase<Vector2D<T>>.IsZero(Vector2D<T> value);
+    static Vector2D<T> INumberBase<Vector2D<T>>.MaxMagnitude(Vector2D<T> x, Vector2D<T> y);
+    static Vector2D<T> INumberBase<Vector2D<T>>.MaxMagnitudeNumber(Vector2D<T> x, Vector2D<T> y);
+    static Vector2D<T> INumberBase<Vector2D<T>>.MinMagnitude(Vector2D<T> x, Vector2D<T> y);
+    static Vector2D<T> INumberBase<Vector2D<T>>.MinMagnitudeNumber(Vector2D<T> x, Vector2D<T> y);
+    static bool INumberBase<Vector2D<T>>.TryConvertFromChecked<TOther>(TOther value, out Vector2D<T> result) where TOther : INumberBase<TOther>;
+    static bool INumberBase<Vector2D<T>>.TryConvertFromSaturating<TOther>(TOther value, out Vector2D<T> result) where TOther : INumberBase<TOther>;
+    static bool INumberBase<Vector2D<T>>.TryConvertFromTruncating<TOther>(TOther value, out Vector2D<T> result) where TOther : INumberBase<TOther>;
+    static bool INumberBase<Vector2D<T>>.TryConvertToChecked<TOther>(Vector2D<T> value, [MaybeNullWhen(false)] out TOther result) where TOther : INumberBase<TOther>;
+    static bool INumberBase<Vector2D<T>>.TryConvertToSaturating<TOther>(Vector2D<T> value, [MaybeNullWhen(false)] out TOther result) where TOther : INumberBase<TOther>;
+    static bool INumberBase<Vector2D<T>>.TryConvertToTruncating<TOther>(Vector2D<T> value, [MaybeNullWhen(false)]out TOther result) where TOther : INumberBase<TOther>;
+
+    // For internal use only, not a public API
+    // Vector2D<T1>? IVec2.GetChecked<T1>();
+    // Vector2D<T1>? IVec2.GetSaturating<T1>();
+    // Vector2D<T1>? IVec2.GetTruncating<T1>();
+
+    // IReadOnlyList<T>
+    IEnumerator<T> IEnumerable<T>.GetEnumerator();
+    IEnumerator IEnumerable.GetEnumerator();
+    int IReadOnlyCollection<T>.Count => Count;
+
+    // IUtf8SpanFormattable, IUtf8SpanParsable<Vector2D<T>>
+    public bool TryFormat(Span<byte> utf8Destination, out int bytesWritten, ReadOnlySpan<char> format, IFormatProvider? provider);
+    public static Vector2D<T> Parse(ReadOnlySpan<byte> utf8Text, IFormatProvider? provider);
+    public static bool TryParse(ReadOnlySpan<byte> s, IFormatProvider? provider, out Vector2D<T> result);
+
+    // IVector<Vector2D<T>>
+    T IVector<Vector2D<T>, T>.LengthSquared();
```

Comment:
nit: `LengthSquared` is effectively a property, not a method.

We're going to be making it a property when the BCL eventually exposes `Vector2<T>`

Reply:
That seems like an uncontroversial change.

Reply:
Update: It occurs to me that we can't add extension properties, and the pre-established convention is for everything possible to be on the non-generic VectorXD types. This would require making an exception for LengthSquared and putting that in VectorXD&lt;T>. We could keep the pattern and have an explicitly implemented IVector property instead, but that makes things unnecessarily inconsistent. CC @Perksey since I believe the design was your suggestion